clients/src/main/java/org/apache/kafka/common/utils/Timer.java

```diff
@@ -0,0 +1,159 @@
+ * This class also ensures monotonic updates to the timer even if the underlying clock is subject
+ * to non-monotonic behavior. For example, the remaining time returned by {@link #remainingMs()} is
+ * guaranteed to decrease monotonically until it hits zero.
+ *
+ * Note that it is up to the caller to ensure progress of the timer using one of the
+ * {@link #update()} methods or {@link #sleep(long)}. The timer will cache the current time and
+ * return it indefinitely until the timer has been updated. This allows the caller to limit
+ * unnecessary system calls and update the timer only when needed. For example, a timer which is
+ * waiting a request sent through the {@link org.apache.kafka.clients.NetworkClient} should call
+ * {@link #update()} following each blocking call to
+ * {@link org.apache.kafka.clients.NetworkClient#poll(long, long)}.
+ *
+ * A typical usage might look something like this:
+ *
+ * <pre>
+ *     Time time = Time.SYSTEM;
+ *     Timer timer = time.timer(500);
+ *
+ *     while (!conditionSatisfied() && !timer.isExpired) {
+ *         client.poll(timer.remainingMs(), timer.currentTimeMs());
+ *         timer.update();
+ *     }
+ * </pre>
+ */
+public class Timer {
+    private final Time time;
+    private long startMs;
+    private long currentTimeMs;
+    private long deadlineMs;
+
+    Timer(Time time, long timeoutMs) {
+        this.time = time;
+        reset(timeoutMs);
+    }
+
+    /**
+     * Check timer expiration. Like {@link #remainingMs()}, this depends on the current cached
+     * time in milliseconds, which is only updated through one of the {@link #update()} methods
+     * or with {@link #sleep(long)};
+     *
+     * @return true if the timer has expired, false otherwise
+     */
+    public boolean isExpired() {
+        return currentTimeMs >= deadlineMs;
+    }
+
+    /**
+     * Reset the timer to the specific timeout. This will use the underlying {@link #Timer(Time, long)}
+     * implementation to update the current cached time in milliseconds and it will set a new timer
+     * deadline.
+     *
+     * @param timeoutMs The new timeout in milliseconds
+     */
+    public void reset(long timeoutMs) {
+        if (timeoutMs < 0)
+            throw new IllegalArgumentException("Invalid negative timeout " + timeoutMs);
+
+        this.startMs = time.milliseconds();
+        this.currentTimeMs = startMs;
+
+        if (currentTimeMs > Long.MAX_VALUE - timeoutMs)
+            this.deadlineMs = Long.MAX_VALUE;
+        else
+            this.deadlineMs = currentTimeMs + timeoutMs;
+    }
+
+    /**
+     * Use the underlying {@link Time} implementation to update the current cached time. If
+     * the underlying time returns a value which is smaller than the current cached time,
+     * the update will be ignored.
+     */
+    public void update() {
+        this.currentTimeMs = Math.max(time.milliseconds(), this.currentTimeMs);
```

Comment:
It would be better to build something like this on top of `time.nanoseconds()` since it is monotonic. The downside is that it can't be used in places where a real date/time is expected. Would the latter be an issue?

Reply:
At the moment, I've only updated the consumer code to use `Timer`. Lower level code like `NetworkClient` still have APIs which depend on the current time in milliseconds. In a future patch, I think we can change this code to use `Timer` as well. The ultimate goal would be to have a Timer interface which only exposed durations of times and did not have a direct dependence on wall clock time. Then we could safely use `nanoseconds` internally.

Reply:
Is this based on real data or hypothetical? :) I shared a link somewhere that shows that this is not true for environments we care about.

Reply:
Well, I did not do any experiments myself, so I'd say it's hypothetical :P It's mainly from some articles I've read before, but well, they are not authentic either.

Reply:
http://pzemtsov.github.io/2017/07/23/the-slow-currenttimemillis.html

Reply:
@ijuma thanks!